### ISSUE TITLE
Client's property::floating signal called when not needed 

### ISSUE THREAD

**Issue by Glaeqen** [awesomeWM/awesome#2698]
Output of `awesome --version`:

awesome v4.3 (Too long)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

**How to reproduce the issue:**
and
**Actual result:**

1. Connect client's signal _property::floating_ to callback and eg. use notify-send to watch it being called.
2. Spawn new window and perform actions from following points on it (callback called during creation, it seems to be alright)
3. Maximize, unmaximize, fullscreen, unfullscreen - callback called each time when it shouldn't be.
4. Float/unfloat (One **Ctrl-Super-Space** press is enough) - callback called as expected
5. Maximize, unmaximize, fullscreen, unfullscreen - After point 4. no more calls to callback.

**Expected result:**

property::floating signal triggered more consistently? I understand that simple properties can be modified not only during floating/unfloating window but this behavior seems to be straight up wrong. If I'm misunderstanding this, sorry for wasting your time.

**Comment by alfunx** [comment]
The problem is that maximized/fullscreen clients are floating. The [definition of `client.floating`](https://awesomewm.org/apidoc/classes/client.html#client.floating) says

> The client floating state. If the client is part of the tiled layout or free floating.

And maximized/fullscreen clients are not part of the tiled layout.

An easy way to get around that would be to check for the maximized/fullscreen state in the callback:

```lua
client.connect_signal("property::floating", function(c)
    if c.maximized or c.fullscreen then end
    naughty.notify { text = c.floating and "floating" or "not floating" }
end)
```

**Comment by Glaeqen** [comment]
> The problem is that maximized/fullscreen clients are floating.

So why after manual float and unfloat of window callback stops being triggered when maximised of _fullscreened_?


My scenario:

**create window**
**toggle maximize**
- floating -> true
- maximized -> true

**toggle maximize**
- floating -> false
- maximized -> false

**toggle floating**
- floating -> true
- maximised -> false

**toggle floating**
- floating -> false
- maximised -> false

**toggle maximize** _INCONSISTENCY_
- floating -> false
- maximised -> true
 _(callback not triggered so this is my guess)_

**toggle maximize** _INCONSISTENCY_
- floating -> false
- maximised -> false
_(callback not triggered so my guess is:)_

**Comment by alfunx** [comment]
Okay I think I got your issue now, the main thing being

> 5. Maximize, unmaximize, fullscreen, unfullscreen - **After point 4. no more calls to callback.**

I think this is related to *implicitly floating* clients. `property::floating` is emitted only if the floating property is set explicitly. If such a client (with `floating = nil`) is maximized, it is considered to be floating (why though, @psychon?). Try this:

```lua
client.connect_signal("property::maximized", function(c)
    naughty.notify { text = (c.floating and "floating" or "not floating") }
    naughty.notify { text = (c.maximized and "maximized" or "not maximized") }
end)
```

Related commit: e1bad41fc

**Comment by psychon** [comment]
TL;DR: I think the issue is wrong. `property::floating` is always emitted when `c.floating` changes (*), so it is not emitted when not needed. Instead, the actual issue here is that there can be three different states of floating, which can be reached via `c.floating = true`, `c.floating = false`, and `c.floating = nil`. The last one is some implicit state where the floating state is derived from other properties like maximized and fullscreen. No one expects this weird state, but new clients end up in it by default.

@alfunx What exactly is your question here?

How clients can be "implicitly floating", or what? If so, then I gave up digging once I found commit 427679b0facd89cc50d72ca68909783b27a608bd from 2008 which added a common `client_isfloating` to the C code, consolidating something that was previously spread around everywhere.

Or where this tristate system of "floating can be set explicitly to true/false or it can be in a default state where it updated based on other properties like maximized" comes from? If so, my digging found commit c3c20c4f8e8e76e4108e83f71bebdec135784c84 which is also from 2008 and which moves handling of the floating state from C to Lua. This commit introduced the `if data.floating[c] ~= nil` which is where this tristate system comes from. I guess this was not done explicitly, but the resulting behaviour was preserved over the years.

> `property::floating` is emitted only if the floating property is set explicitly.

No. It is emitted when the floating property of a client changes. If I understand the initial comment in here correctly, it is basically "this tristate system makes no sense".

A bit more explicit:

> * Connect client's signal _property::floating_ to callback and eg. use notify-send to watch it being called.
> * Spawn new window and perform actions from following points on it (callback called during creation, it seems to be alright)

Let's assume the default: The window ends up with `c.floating == false` in this situation.

> * Maximize, unmaximize, fullscreen, unfullscreen - callback called each time when it shouldn't be.

If you look at `c.floating`, it will be `true` in this situation. Thus, the floating state of the client actually changed. Thus, it makes sense that the signal is emitted. This should e.g. be visible in the icon indicating the floating state in the titlebar. (**)

> * Float/unfloat (One **Ctrl-Super-Space** press is enough) - callback called as expected

Okay.

> * Maximize, unmaximize, fullscreen, unfullscreen - After point 4. no more calls to callback.

Well, at this point the client was assigned an explicit floating state via `awful.client.object.set_floating`. Thus, the implicit default is no longer applied.

(*): Well, I can imagine one situation where it is wrongly emitted: If a client is currently implicitly floating because it is maximized and you do `c.floating = true`, you should get a `property::floating` even though `c.floating` did not actually change. However, I also think that this case is not discussed in the other comments in here.
(**): Indeed. Open a new window and maximize it with Mod4+M. Both the maximized and the floating icon change. Do this again and make the window floating, then unfloating. Now, Mod4+M only changes the maximized state.

**Comment by alfunx** [comment]
Sorry I got confused there, this explains it for me

> The last one [`c.floating = nil`] is some implicit state where the floating state is derived from other properties like maximized and fullscreen.

In that case, this sounds like a non-issue, since it's a convention. For OP, I guess setting `floating = false` (through a rule for example) would be a valid solution, if you have a problem with that behavior.
